# progeny-py
This is a Python implementation of the package [PROGENy](https://github.com/saezlab/progeny) for pathway activity prediction from transcriptomics data.
In this implementation, PROGENy uses the [scanpy](https://github.com/theislab/scanpy) framework. 


## Installation
The package can easly be installed using pip
```
pip install progeny-py
```

## Examples
Jupyter notebooks explaining the method and the package can be found in the `example/` directory


## Citing PROGENy

Besides the original paper, there are two additional publication 
describing expansions of PROGENy usage.

- If you use PROGENy for your research please cite the original publication:

> Schubert M, Klinger B, Klünemann M, Sieber A, Uhlitz F, Sauer S, Garnett MJ, Blüthgen N, Saez-Rodriguez J. “Perturbation-response genes reveal signaling footprints in cancer gene expression.” _Nature Communications_: [10.1038/s41467-017-02391-6](https://doi.org/10.1038/s41467-017-02391-6)

- If you use for mouse or you use the expanded version containing 14 pathways, 
please cite additionally:

> Holland CH, Szalai B, Saez-Rodriguez J. "Transfer of regulatory knowledge from human to mouse for functional genomics analysis." _Biochimica et Biophysica Acta (BBA) - Gene Regulatory Mechanisms._ 2019. DOI: [10.1016/j.bbagrm.2019.194431](https://doi.org/10.1016/j.bbagrm.2019.194431).

- If you apply PROGENy on single-cell RNA-seq data please cite additionally:

> Holland CH, Tanevski J, Perales-Patón J, Gleixner J, Kumar MP, Mereu E, Joughin BA, Stegle O, Lauffenburger DA, Heyn H, Szalai B, Saez-Rodriguez, J. "Robustness and applicability of transcription factor and pathway analysis tools on single-cell RNA-seq data." _Genome Biology._ 2020. DOI: [10.1186/s13059-020-1949-z](https://doi.org/10.1186/s13059-020-1949-z).